packages/shared-crypto/lib/keyring.ts

```diff
@@ -0,0 +1,314 @@
+/*
+Copyright 2018 - 2024 The Alephium Authors
+This file is part of the alephium project.
+
+The library is free software: you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+The library is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License
+along with the library. If not, see <http://www.gnu.org/licenses/>.
+*/
+
+import { AddressHash } from '@alephium/shared'
+import { addressToGroup, bs58, ExplorerProvider, sign, TOTAL_NUMBER_OF_GROUPS, transactionSign } from '@alephium/web3'
+import * as bip32 from 'bip32'
+import * as bip39 from 'bip39'
+import blake from 'blakejs'
+
+import { findNextAvailableAddressIndex, isAddressIndexValid } from './address'
+import { decryptMnemonic, MnemonicLength } from './wallet'
+
+export type NonSensitiveAddressData = {
+  hash: AddressHash
+  index: number
+  publicKey: string
+}
+
+type SensitiveAddressData = NonSensitiveAddressData & {
+  privateKey: Buffer
+}
+
+type GenerateAddressProps = {
+  group?: number
+  addressIndex?: number
+  skipAddressIndexes?: number[]
+}
+
+class Keyring {
+  private hdPath = "m/44'/1234'/0'/0"
+  private root: bip32.BIP32Interface | null
+  private addresses: SensitiveAddressData[]
```

Comment:
This corresponds to this: https://github.com/MetaMask/eth-hd-keyring/blob/main/index.js#L32